pilot/pkg/serviceregistry/kube/controller/ambientindex_external.go

```diff
@@ -313,3 +496,118 @@ func (c *Controller) getControllerWorkloadEntries(ns string) []*apiv1alpha3.Work
 	}
 	return allWorkloadEntries
 }
+
+// name format: <cluster>/<group>/<kind>/<namespace>/<name></section-name>
+// if the WorkloadEntry is inlined in the ServiceEntry, we may need section name. caller should use generateServiceEntryUID
+func (c *Controller) generateWorkloadEntryUID(wkEntryNamespace, wkEntryName string) string {
+	return c.clusterID.String() + "/networking.istio.io/WorkloadEntry/" + wkEntryNamespace + "/" + wkEntryName
+}
+
+// name format: <cluster>/<group>/<kind>/<namespace>/<name></section-name>
+// section name should be the WE address, which needs to be stable across SE updates (it is assumed WE addresses are unique)
+func (c *Controller) generateServiceEntryUID(svcEntryNamespace, svcEntryName, addr string) string {
+	return c.clusterID.String() + "/networking.istio.io/ServiceEntry/" + svcEntryNamespace + "/" + svcEntryName + "/" + addr
+}
+
+func (a *AmbientIndexImpl) cleanupOldWorkloadEntriesInlinedOnServiceEntry(svc *model.Service, updates sets.Set[model.ConfigKey], c *Controller) {
+	nsName := types.NamespacedName{
+		Name:      svc.Attributes.ServiceEntryName,
+		Namespace: svc.Attributes.ServiceEntryNamespace,
+	}
+
+	// cleanup any old WorkloadEntries generated from this ServiceEntry (`ServiceEntry.endpoints`)
+	if oldServiceEntry, f := a.servicesMap[nsName]; f && oldServiceEntry.Attributes.ServiceEntry != nil {
+		for _, oldWe := range oldServiceEntry.Attributes.ServiceEntry.Endpoints {
+			oldUID := c.generateServiceEntryUID(nsName.Namespace, nsName.Name, oldWe.Address)
+			we, found := a.byUID[oldUID]
+			if found {
+				updates.Insert(model.ConfigKey{Kind: kind.Address, Name: we.ResourceName()})
+				for _, networkAddr := range networkAddressFromWorkload(we) {
+					delete(a.byWorkloadEntry, networkAddr)
+				}
+				delete(a.byUID, oldUID)
+			}
+		}
+	}
+}
+
+func getWorkloadServices(serviceEntries map[types.NamespacedName]*model.Service, workloadEntry *v1alpha3.WorkloadEntry,
```

Comment:
nit: `getWorkloadServiceEntries` may be more clear?

Reply:
addressed in https://github.com/istio/istio/pull/45621/commits/04b74353533953c046cf05f281d64e4c55a3a736